src/main/java/com/google/devtools/build/lib/remote/TreeNodeRepository.java

```diff
@@ -362,23 +362,21 @@ private synchronized Directory getOrComputeDirectory(TreeNode node) throws IOExc
         TreeNode child = entry.getChild();
         if (child.isLeaf()) {
           ActionInput input = child.getActionInput();
+          final Digest digest;
           if (input instanceof VirtualActionInput) {
             VirtualActionInput virtualInput = (VirtualActionInput) input;
-            Digest digest = digestUtil.compute(virtualInput);
+            digest = digestUtil.compute(virtualInput);
             virtualInputDigestCache.put(virtualInput, digest);
             // There may be multiple inputs with the same digest. In that case, we don't care which
             // one we get back from the digestVirtualInputCache later.
             digestVirtualInputCache.put(digest, virtualInput);
-            b.addFilesBuilder()
-                .setName(entry.getSegment())
-                .setDigest(digest)
-                .setIsExecutable(false);
           } else {
-            b.addFilesBuilder()
-                .setName(entry.getSegment())
-                .setDigest(DigestUtil.getFromInputCache(input, inputFileCache))
-                .setIsExecutable(execRoot.getRelative(input.getExecPathString()).isExecutable());
+            digest = DigestUtil.getFromInputCache(input, inputFileCache);
           }
+          b.addFilesBuilder()
+              .setName(entry.getSegment())
+              .setDigest(digest)
+              .setIsExecutable(true);
```

Comment:
I need help understanding this - instead of changing this in RemoteSpawnRunner, where the Remote Ex Action is built, you want to change this here, even where it doesn't relate to files specifically, but so that it can be treated in the a) actionKey computation and b) remote execution side as executable, and c) really so that osx can avoid an additional stat of an input, or that a mismatch causes an effective `chmod +x` ?

Does this bug not appear in standalone? sandbox? I have so many questions...

Reply:
This code is being used in the RemoteSpawnRunner and the RemoteSpawnCache. The whole point of the change is to not to the stat() call which for some miraculous reason fails sometimes on macOS. We have never seen this bug appear in any other spawn strategy.

Reply:
> It's your code at the end of the day, but I would much rather see the effort expended to route out the unsafe file operations than 'avoid a stat' to workaround either a misuse of files, a JVM bug, or an OS race.

Me too. We currently don't have anyone with the time to look into it and if this buys us some time, I am happy to be pragmatic and unblock users.

Reply:
I am happy to try it out. Can you please send me an e-mail with instructions? I'll likely overlook Github comments :-).